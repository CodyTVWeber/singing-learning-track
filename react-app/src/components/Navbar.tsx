import React, { useState } from 'react';
import { colors, spacing, shadows, transitions, fontSize, fontWeight, borderRadius } from '../theme/theme';

interface NavLink {
  label: string;
  href: string;
  icon?: React.ReactNode;
  active?: boolean;
}

interface NavbarProps {
  brand?: React.ReactNode;
  links?: NavLink[];
  onLinkClick?: (link: NavLink) => void;
  sticky?: boolean;
  transparent?: boolean;
  className?: string;
  style?: React.CSSProperties;
}

export const Navbar: React.FC<NavbarProps> = ({
  brand,
  links = [],
  onLinkClick,
  sticky = true,
  transparent = false,
  className = '',
  style,
}) => {
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);

  const navbarStyles: React.CSSProperties = {
    position: sticky ? 'sticky' : 'relative',
    top: 0,
    left: 0,
    right: 0,
    backgroundColor: transparent ? 'transparent' : colors.surface,
    boxShadow: transparent ? 'none' : shadows.sm,
    transition: transitions.normal,
    zIndex: 1000,
    ...style,
  };

  const containerStyles: React.CSSProperties = {
    maxWidth: '1200px',
    margin: '0 auto',
    padding: `${spacing.md} ${spacing.lg}`,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  };

  const brandStyles: React.CSSProperties = {
    fontSize: fontSize.xl,
    fontWeight: fontWeight.bold,
    color: colors.primary,
    display: 'flex',
    alignItems: 'center',
    gap: spacing.sm,
  };

  const desktopLinksStyles: React.CSSProperties = {
    display: 'flex',
    alignItems: 'center',
    gap: spacing.lg,
    '@media (max-width: 768px)': {
      display: 'none',
    },
  };

  const linkStyles = (active?: boolean): React.CSSProperties => ({
    display: 'flex',
    alignItems: 'center',
    gap: spacing.xs,
    padding: `${spacing.sm} ${spacing.md}`,
    borderRadius: borderRadius.md,
    fontSize: fontSize.md,
    fontWeight: active ? fontWeight.semibold : fontWeight.normal,
    color: active ? colors.primary : colors.text,
    backgroundColor: active ? colors.featherLight : 'transparent',
    transition: transitions.fast,
    cursor: 'pointer',
  });

  const mobileMenuButtonStyles: React.CSSProperties = {
    display: 'none',
    '@media (max-width: 768px)': {
      display: 'flex',
    },
    width: '40px',
    height: '40px',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: borderRadius.md,
    backgroundColor: 'transparent',
    cursor: 'pointer',
    transition: transitions.fast,
  };

  const mobileMenuStyles: React.CSSProperties = {
    position: 'absolute',
    top: '100%',
    left: 0,
    right: 0,
    backgroundColor: colors.surface,
    boxShadow: shadows.lg,
    maxHeight: mobileMenuOpen ? '400px' : '0',
    overflow: 'hidden',
    transition: transitions.normal,
  };

  const mobileLinksContainerStyles: React.CSSProperties = {
    padding: spacing.md,
    display: 'flex',
    flexDirection: 'column',
    gap: spacing.xs,
  };

  const hamburgerLineStyles: React.CSSProperties = {
    width: '24px',
    height: '2px',
    backgroundColor: colors.text,
    transition: transitions.fast,
  };

  return (
    <nav className={className} style={navbarStyles}>
      <div style={containerStyles}>
        {brand && <div style={brandStyles}>{brand}</div>}
        
        {/* Desktop Links */}
        <div style={desktopLinksStyles}>
          {links.map((link, index) => (
            <a
              key={index}
              href={link.href}
              style={linkStyles(link.active)}
              onClick={(e) => {
                e.preventDefault();
                onLinkClick?.(link);
              }}
              onMouseEnter={(e) => {
                if (!link.active) {
                  e.currentTarget.style.backgroundColor = colors.background;
                }
              }}
              onMouseLeave={(e) => {
                if (!link.active) {
                  e.currentTarget.style.backgroundColor = 'transparent';
                }
              }}
            >
              {link.icon}
              {link.label}
            </a>
          ))}
        </div>

        {/* Mobile Menu Button */}
        <button
          style={mobileMenuButtonStyles}
          onClick={() => setMobileMenuOpen(!mobileMenuOpen)}
          aria-label="Toggle mobile menu"
        >
          <div style={{ display: 'flex', flexDirection: 'column', gap: '4px' }}>
            <div style={{
              ...hamburgerLineStyles,
              transform: mobileMenuOpen ? 'rotate(45deg) translateY(6px)' : 'none',
            }} />
            <div style={{
              ...hamburgerLineStyles,
              opacity: mobileMenuOpen ? 0 : 1,
            }} />
            <div style={{
              ...hamburgerLineStyles,
              transform: mobileMenuOpen ? 'rotate(-45deg) translateY(-6px)' : 'none',
            }} />
          </div>
        </button>
      </div>

      {/* Mobile Menu */}
      <div style={mobileMenuStyles}>
        <div style={mobileLinksContainerStyles}>
          {links.map((link, index) => (
            <a
              key={index}
              href={link.href}
              style={linkStyles(link.active)}
              onClick={(e) => {
                e.preventDefault();
                setMobileMenuOpen(false);
                onLinkClick?.(link);
              }}
            >
              {link.icon}
              {link.label}
            </a>
          ))}
        </div>
      </div>

      <style>{`
        @media (max-width: 768px) {
          .navbar-desktop-links { display: none !important; }
          .navbar-mobile-button { display: flex !important; }
        }
        @media (min-width: 769px) {
          .navbar-desktop-links { display: flex !important; }
          .navbar-mobile-button { display: none !important; }
        }
      `}</style>
    </nav>
  );
};