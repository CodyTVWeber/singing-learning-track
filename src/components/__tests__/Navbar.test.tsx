import React from 'react';
import { render } from '@testing-library/react';
import { Navbar } from '../Navbar';

describe('Navbar', () => {
  it('renders with links snapshot', () => {
    const { asFragment } = render(
      <Navbar
        brand={"Kooka"}
        links={[
          { label: 'Home', href: '#', active: true },
          { label: 'About', href: '#' },
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

