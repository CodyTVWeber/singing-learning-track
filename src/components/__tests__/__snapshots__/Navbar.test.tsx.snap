// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > renders with links snapshot 1`] = `
<DocumentFragment>
  <nav
    class=""
    style="position: sticky; top: 0px; left: 0px; right: 0px; background-color: rgb(255, 255, 255); box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05); transition: 300ms ease-in-out; z-index: 1000;"
  >
    <div
      style="max-width: 1200px; margin: 0px auto; padding: 16px 24px; display: flex; align-items: center; justify-content: space-between;"
    >
      <div
        style="font-size: 24px; font-weight: 700; color: rgb(139, 105, 82); display: flex; align-items: center; gap: 8px;"
      >
        Kooka
      </div>
      <div
        class="navbar-desktop-links"
        style="display: flex; align-items: center; gap: 24px;"
      >
        <a
          href="#"
          style="display: flex; align-items: center; gap: 4px; padding: 8px 16px; border-radius: 8px; font-size: 16px; font-weight: 600; color: rgb(139, 105, 82); background-color: rgb(215, 204, 200); transition: 150ms ease-in-out; cursor: pointer;"
        >
          Home
        </a>
        <a
          href="#"
          style="display: flex; align-items: center; gap: 4px; padding: 8px 16px; border-radius: 8px; font-size: 16px; font-weight: 400; color: rgb(62, 39, 35); background-color: transparent; transition: 150ms ease-in-out; cursor: pointer;"
        >
          About
        </a>
      </div>
      <button
        aria-label="Toggle mobile menu"
        class="navbar-mobile-button"
        style="display: flex; width: 40px; height: 40px; align-items: center; justify-content: center; border-radius: 8px; background-color: transparent; cursor: pointer; transition: 150ms ease-in-out;"
      >
        <div
          style="display: flex; flex-direction: column; gap: 4px;"
        >
          <div
            style="width: 24px; height: 2px; background-color: rgb(62, 39, 35); transition: 150ms ease-in-out; transform: none;"
          />
          <div
            style="width: 24px; height: 2px; background-color: rgb(62, 39, 35); transition: 150ms ease-in-out; opacity: 1;"
          />
          <div
            style="width: 24px; height: 2px; background-color: rgb(62, 39, 35); transition: 150ms ease-in-out; transform: none;"
          />
        </div>
      </button>
    </div>
    <div
      style="position: absolute; top: 100%; left: 0px; right: 0px; background-color: rgb(255, 255, 255); box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05); max-height: 0; overflow: hidden; transition: 300ms ease-in-out;"
    >
      <div
        style="padding: 16px; display: flex; flex-direction: column; gap: 4px;"
      >
        <a
          href="#"
          style="display: flex; align-items: center; gap: 4px; padding: 8px 16px; border-radius: 8px; font-size: 16px; font-weight: 600; color: rgb(139, 105, 82); background-color: rgb(215, 204, 200); transition: 150ms ease-in-out; cursor: pointer;"
        >
          Home
        </a>
        <a
          href="#"
          style="display: flex; align-items: center; gap: 4px; padding: 8px 16px; border-radius: 8px; font-size: 16px; font-weight: 400; color: rgb(62, 39, 35); background-color: transparent; transition: 150ms ease-in-out; cursor: pointer;"
        >
          About
        </a>
      </div>
    </div>
    <style>
      
        @media (max-width: 768px) {
          .navbar-desktop-links { display: none !important; }
          .navbar-mobile-button { display: flex !important; }
        }
        @media (min-width: 769px) {
          .navbar-desktop-links { display: flex !important; }
          .navbar-mobile-button { display: none !important; }
        }
      
    </style>
  </nav>
</DocumentFragment>
`;
